feat: Implement kubernetes recommended labels

samuel-hawker opened the pull request:
Implement the kubernetes recommended labels:
app.kubernetes.io/name
app.kubernetes.io/instance
app.kubernetes.io/managed-by
Applies for all operators
Each resource is labelled with which operator
component it is managed by

Signed-off-by: Samuel Hawker <samuel.hawker@ibm.com>

### Type of change

_Select the type of your PR_

- Bugfix
- Enhancement / new feature
- Refactoring
- Documentation

### Description

_Please describe your pull request_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityOperator.java:
```diff
@@ -62,7 +64,7 @@
      */
     protected EntityOperator(String namespace, String cluster, Labels labels) {
-        super(namespace, cluster, labels);
+        super(namespace, cluster, labels, ENTITY_OPERATOR_NAME);
```
I'm not sure this is correct. The Entity Operator is managed by the Cluster Operator. It doesn't manage it self. So I think everything created through the Abstract model should have the _managed by_ label set to something like `strimzi-cluster-operator`. I this applies to most of the files in the `cluster-operator` module. Do we really want/need to distinguish it beyond the process level? I think the process level (i.e. which pod is managing it might be more understandable for the user).

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java:
```diff
@@ -178,13 +178,17 @@
-     * @param cluster   overall cluster name
+     * @param namespace    Kubernetes/OpenShift namespace where cluster resources are going to be created
+     * @param cluster      overall cluster name
+     * @param operatorName overall cluster name
```
it's "operator name"

ppatierno commented on operator-common/src/main/java/io/strimzi/operator/common/model/Labels.java:
```diff
@@ -178,6 +186,33 @@ public Labels withKind(String kind) {
+    
+    /**
+     * The same labels as this instance, but with the application name {@code strimzi} for the {@code app.kubernetes.io/name} key.
+     * @return A new instance with the given kubernetes kubernetes application name added.
```
double "kubernetes"

samuel-hawker commented:
@scholzj I have changed all references of the managed-by key to use the `strimzi-cluster-operator` value.
I am currently looking into applying similar labels for the topic and user operators... but having some issues with getting the tests working.

@ppatierno I have fixed all of the typos :)

Would you be open to accepting these changes incrementally - i.e. I apply the labels for topic and user operators in a followup PR, or would you prefer all changes in the same set of deltas?

ppatierno commented:
@samuel-hawker I am not a big fan of big PRs because they are hard to review. Anyway, I think that in this case, the changes for user and topic operators should not be so big. I guess that they can fit in this PR. But I am fine with a new one if  @scholzj prefers that.

scholzj commented:
I'm totally fine to merge this PR and have another one for Topic / User operator.

scholzj commented:
@strimzi-ci run tests

scholzj commented:
@strimzi-ci run tests
testcase=io.strimzi.systemtest.RecoveryST#io.strimzi.systemtest.RecoveryST,testKafkaAndZookeeperScaleUpScaleDown#io.strimzi.systemtest.KafkaST,io.strimzi.systemtest.AllNamespaceST#io.strimzi.systemtest.AllNamespaceST,testMirrorMakerTlsAuthenticated#io.strimzi.systemtest.MirrorMakerST

scholzj commented:
@strimzi-ci run tests
testcase=io.strimzi.systemtest.RecoveryST#io.strimzi.systemtest.RecoveryST,testKafkaAndZookeeperScaleUpScaleDown#io.strimzi.systemtest.KafkaST,io.strimzi.systemtest.AllNamespaceST#io.strimzi.systemtest.AllNamespaceST,testMirrorMakerTlsAuthenticated#io.strimzi.systemtest.MirrorMakerST

ppatierno commented on operator-common/src/main/java/io/strimzi/operator/common/model/Labels.java:
```diff
@@ -178,6 +186,33 @@ public Labels withKind(String kind) {
+     * @return A new instance with the given operator that is managing this resourse.
+     */
+    public Labels withKubernetesManagedBy(String operatorName) {
+        // Make configurable?
```
what does this comment mean? should we remove it, or it raises idea for an improvement?

samuel-hawker replied:
Lol, that was mid-development comment. From when I thought that the `Labels` class was used by all the operators (including user and topic operator).
Since user and topic operators have their own implementations of labels this does not need to be configurable, but it does make me wonder, is there any reason why we can't pull the Labels class into operator-common and have all three operator types use the same labels class/utilities?

samuel-hawker replied:
I can remove it in this PR, or I can remove it in my followup user/topic operator changes.

Delta can be found here if interested: https://github.com/strimzi/strimzi-kafka-operator/commit/f853383080fa6baa814ad706546780f86b08f5fc

Hoping to get it finished soon, but struggling with the Topic Operator as the session tests keep timing out when I add the labels!

ppatierno replied:
Good point about having the `Labels` class common across operators. For now, it's enough just removing the comment and we are ready to merge :-)

scholzj commented:
Thaanks for the PR. I will merge it now.
